### PR TITLE
Contract deployment QoL

### LINKED ISSUE #217: Deployment and loading of contract QoL can be improved
Currently, deploying and loading contracts is done in multiple steps, but could be done more ergonomically.

As per my issue:
https://github.com/FuelLabs/fuels-rs/issues/215#issuecomment-1101622119

```rust
use fuels_abigen_macro::abigen;
use fuels_contract::contract::Contract;

// Generate Rust bindings from our contract JSON ABI
abigen!(MyContract, "./my-contract-abi.json");

let compiled = Contract::load_sway_contract(path,salt).unwrap();
let id = Contract::deploy(&compiled, &provider, &wallet, TxParameters::default())
        .await
        .unwrap();
```

Could just return a contract instance for example and a helper function could do the two activities (as they are often done in tandem anyway):

```rust
use fuels_abigen_macro::abigen;
use fuels_contract::contract::Contract;

// Generate Rust bindings from our contract JSON ABI
abigen!(MyContract, "./my-contract-abi.json");

let contract_instance = Contract::deploy_compiled(path, MyContract, &wallet, TxParameters::default())
        .await
        .unwrap();
```

Returning the contract instance is far more helpful. Including RNG generation is also nice. Wallet like Ethers.js / RS, could include a provider property to grab.

As well, MyContract could also have the path to the contracts included at that stage, i.e.:

```rust
abigen!(MyContract, "./my-contract-abi.json", "./");

let contract_instance = Contract::deploy_compiled(MyContract, &wallet, TxParameters::default())
        .await
        .unwrap();
```

I think approaches with the mentality above could make everything even more ergonomic allowing the developer to focus on just the logic they need to think about and less config and wiring that should just be inherited.

### PR DESCRIPTION
**Pull request by MujkicA**
Closes #217

### CODE REVIEW

File: packages/fuels-core/src/code_gen/abigen.rs [hunk 1]
```diff
@@ -36,7 +38,11 @@ pub struct Abigen {
 
 impl Abigen {
     /// Creates a new contract with the given ABI JSON source.
-    pub fn new<S: AsRef<str>>(contract_name: &str, abi_source: S) -> Result<Self, Error> {
+    pub fn new<S: AsRef<str>>(
+        contract_name: &str,
+        abi_source: S,
+        binary_filepath: &Option<String>,
```
Comment (leviathanbeak) [inline packages/fuels-core/src/code_gen/abigen.rs:44]: any special reason this isn't `Option<&str>` instead of `&Option<String>` ?

File: packages/fuels-abigen-macro/Cargo.toml [hunk 2]
```diff
@@ -12,6 +12,7 @@ description = "Fuel Rust SDK marcros to generate types from ABI."
 proc-macro = true
 
 [dependencies]
+async-trait = "0.1.53"
```
Comment (leviathanbeak) [inline packages/fuels-abigen-macro/Cargo.toml:15]: `0.1` should suffice

File: packages/fuels-core/src/code_gen/abigen.rs (hunk 1 above)
Comment (MujkicA) [reply packages/fuels-core/src/code_gen/abigen.rs]: adjusted it

File: packages/fuels-contract/src/contract.rs [hunk 3]
```diff
@@ -571,14 +596,14 @@ where
 
     /// Call a contract's method on the node, in a state-modifying manner.
     pub async fn call(self) -> Result<CallResponse<D>, Error> {
-        Ok(Self::call_or_simulate(self, false).await?)
+        Self::call_or_simulate(self, false).await
```
Comment (vnepveu) [inline packages/fuels-contract/src/contract.rs:599]: why did you remove the `?` ? Maybe I don't understand how the `?` works haha

File: packages/fuels-contract/src/contract.rs [hunk 4]
```diff
@@ -22,6 +23,12 @@ use fuels_signers::provider::Provider;
 use fuels_signers::{LocalWallet, Signer};
 use std::marker::PhantomData;
 
+#[async_trait]
+pub trait Deployable {
```
Comment (vnepveu) [inline packages/fuels-contract/src/contract.rs:27]: I don't understand the point of this trait if it is used only once. Traits are used, afaik, to have a sort of generic interface. In our case, only `Contracts` are `Deployable` basically, so it does not make sense IMO to create a whole trait for it, especially if you have to use a new crate for the trait to exist.

File: packages/fuels-abigen-macro/src/lib.rs [hunk 5]
```diff
@@ -21,7 +26,13 @@ pub fn abigen(input: TokenStream) -> TokenStream {
 pub fn wasm_abigen(input: TokenStream) -> TokenStream {
     let args = parse_macro_input!(input as Spanned<ContractArgs>);
 
-    let c = Abigen::new(&args.name, &args.abi).unwrap().no_std();
+    let c = Abigen::new(
+        &args.name,
+        &args.abi,
+        args.binary_filepath.as_ref().map(|v| v.as_ref()),
```
Comment (vnepveu) [inline packages/fuels-abigen-macro/src/lib.rs:32]: not a big fan of this `.map()` call, it seems weird to me

File: packages/fuels-abigen-macro/src/lib.rs [hunk 6]
```diff
@@ -90,10 +102,28 @@ impl ParseInner for ContractArgs {
             let literal = input.parse::<LitStr>()?;
             (literal.span(), literal.value())
         };
+
+        if !input.is_empty() {
+            input.parse::<Token![,]>()?;
+        }
+
+        let binary_filepath = if !input.is_empty() {
```
Comment (vnepveu) [inline packages/fuels-abigen-macro/src/lib.rs:110]: it should probably be refactored with the statement below IMO

File: packages/fuels-abigen-macro/tests/harness.rs [hunk 7]
```diff
@@ -608,10 +608,10 @@ async fn create_nested_struct_from_decoded_tokens() {
     assert_eq!(10, nested_struct_from_tokens.x);
     assert!(nested_struct_from_tokens.y.a);
 
-    let (provider, wallet) = setup_test_provider_and_wallet().await;
+    let (_, wallet) = setup_test_provider_and_wallet().await;
```
Comment (vnepveu) [inline packages/fuels-abigen-macro/tests/harness.rs:611]: the fact that the provider is consistently put as an ignored variable (`_`) tells me that either:
- the function  should be refactored
- a new function should be written

File: packages/fuels-core/src/code_gen/abigen.rs [hunk 8]
```diff
@@ -335,7 +381,34 @@ mod tests {
...
         // there are 5
-        let _bindings = Abigen::new("test", contract).unwrap().generate().unwrap();
+        let _bindings = Abigen::new("test", contract, None)
+            .unwrap()
+            .generate()
+            .unwrap();
+    }
+
+    #[test]
+    fn generates_deployable() {
```
Comment (vnepveu) [inline packages/fuels-core/src/code_gen/abigen.rs:391]: This test does not test anything from what I understand, you should probably add an assert statement or expect a panic from the test

File: packages/fuels-core/src/code_gen/abigen.rs [hunk 9]
```diff
@@ -152,13 +166,42 @@ impl Abigen {
...
                 #code
+                #deployable_trait
 
                 #abi_structs
                 #abi_enums
             }
         })
     }
 
+    fn deployable_trait(&self, name: &Ident) -> TokenStream {
```
Comment (vnepveu) [inline packages/fuels-core/src/code_gen/abigen.rs:177]: The function should be name `generate_deployable_trait` or similar. I'm not sure of the relevance of it though.

File: packages/fuels-core/src/code_gen/abigen.rs [hunk 10]
```diff
@@ -152,13 +166,42 @@ impl Abigen {
...
+        if let Some(binary_filepath) = self.binary_filepath.as_ref() {
+            return quote! {
+                #[async_trait]
+                impl Deployable for #name {
+                    async fn deploy_compiled(wallet: LocalWallet, params: TxParameters) -> Self {
```
Comment (vnepveu) [inline packages/fuels-core/src/code_gen/abigen.rs:182]: The function should not be called `deploy_compiled` but rather `compile_and_deploy`. Also this function should IMO not be in the ABIgen, it is a function that should be handled on the SDK side (like for instance the `load_sway` or `compile` function) and should not be generated as part of a macro call.

**Review by vnepveu — CHANGES_REQUESTED** [review]
I'm having trouble understanding your logic here, looking forward to understand it better!

File: packages/fuels-contract/src/contract.rs (hunk 3 above)
Comment (leviathanbeak) [reply packages/fuels-contract/src/contract.rs:599]: `call_or_simulate()` already returns the same type, so the old code is redundant

File: packages/fuels-core/src/code_gen/abigen.rs (hunk 10 above)
Comment (digorithm) [reply packages/fuels-core/src/code_gen/abigen.rs:182]: > The function should not be called deploy_compiled but rather compile_and_deploy

The SDK isn't responsible for compiling. We are not compiling the Sway code here. We're simply `deploy`ing the `compiled` code, hence `deploy_compiled`.

File: packages/fuels-core/src/code_gen/abigen.rs (hunk 8 above)
Comment (MujkicA) [reply packages/fuels-core/src/code_gen/abigen.rs:391]: The idea was to just have a simple test case which triggers the generation of the `Deployable` implementation for the binding

File: packages/fuels-abigen-macro/tests/harness.rs (hunk 7 above)
Comment (MujkicA) [reply packages/fuels-abigen-macro/tests/harness.rs:611]: I don't have a good enough idea of all the use cases users might have. There are some scenarios where direct interaction with the provider is needed, but it feels awkward do it via the wallet.

File: packages/fuels-contract/src/contract.rs (hunk 4 above)
Comment (MujkicA) [reply packages/fuels-contract/src/contract.rs:27]: I've added some more context as to why this is needed in a comment. Please take a look and let me know what you think.

**Comment by digorithm** [comment]
That's looking really nice. Can we make it look like

```Rust
abigen!(
    MyContract,
    "tests/test_projects/contract_test",
);
```

So that `abigen!` could infer both the `.bin` and `.json`? You mentioned something about this potentially leading to some errors on the user side. Could you expand on that? It would be super nice, from a user perspective, to have to include only a single project path.

**Comment by MujkicA** [comment]
> That's looking really nice. Can we make it look like
> 
> ```rust
> abigen!(
>     MyContract,
>     "tests/test_projects/contract_test",
> );
> ```
> 
> So that `abigen!` could infer both the `.bin` and `.json`? You mentioned something about this potentially leading to some errors on the user side. Could you expand on that? It would be super nice, from a user perspective, to have to include only a single project path.

Agreed, it would look a lot better, however we would need some way to make the user know if they are in deployable mode or if loading the binary failed. 
As for the errors, assuming we fetch files with .json/.bin extensions, what happens if the folder structure does not match the default /out/debug or if there are unrelated files with the same extension. I fear such errors might not be transparent and lead to frustration. Or am I being too cautious here?

File: packages/fuels-contract/src/contract.rs (hunk 4 above)
Comment (MujkicA) [reply packages/fuels-contract/src/contract.rs:27]: Actually you were right, we can do without the trait here. I went trough quite a few iterations and the trait is just a remnant. There is still one advantage to it, if the user defines functions `deploy_compiled()`, `new()`, `get_binary_filepath()` on the sway side, abigen will fail. This way they could coexist and be accessed via fully qualified syntax.

File: packages/fuels-abigen-macro/src/lib.rs (hunk 5 above)
Comment (leviathanbeak) [reply packages/fuels-abigen-macro/src/lib.rs:32]: @vnepveu what's wrong with the map function? do you have any cleaner way to _map_ `Option<T>` to `Option<U>` 🤔

Comment (vnepveu) [reply packages/fuels-abigen-macro/src/lib.rs:32]: I  think I'm just not used to seeing `map` being used on something that is not iterable and only has one element. But if that's the only way then let's go!

File: packages/fuels-abigen-macro/tests/harness.rs (hunk 7 above)
Comment (vnepveu) [reply packages/fuels-abigen-macro/tests/harness.rs:611]: I think having a function that is just `setup_test_wallet` (that the `setup_test_provider_and_wallet` would use) would be good here! This way you only get what you need in the cases where you used `_`, but we can still get the `provider` when it's needed. What do you think?

File: packages/fuels-abigen-macro/src/lib.rs (hunk 5 above)
Comment (leviathanbeak) [reply packages/fuels-abigen-macro/src/lib.rs:32]: it's a different type of `map` the one that works on `Option<T>` types:

https://doc.rust-lang.org/std/option/enum.Option.html#method.map

File: packages/fuels-abigen-macro/tests/harness.rs [hunk 11]
```diff
@@ -1049,34 +977,30 @@ async fn test_large_return_data() {
 async fn test_provider_launch_and_connect() {
```
Comment (vnepveu) [inline packages/fuels-abigen-macro/tests/harness.rs:977]: I think we ought to keep an E2E test for the `load_sway` and `deploy` workflow. Even if it's not the one that we are going to mainly use, we still need to have it tested somewhere. So keep this test and add another one with the "new" workflow (can be in the same test probably)

**Review by vnepveu — CHANGES_REQUESTED** [review]
thank you for the context, much clearer! still some questions

**Comment by digorithm** [comment]
> Agreed, it would look a lot better, however we would need some way to make the user know if they are in deployable mode or if loading the binary failed. 

This is a bit unclear to me. What do you mean by deployable mode?

> As for the errors, assuming we fetch files with .json/.bin extensions, what happens if the folder structure does not match the default /out/debug or if there are unrelated files with the same extension. I fear such errors might not be transparent and lead to frustration. Or am I being too cautious here?

Ah, this is a good point. Then maybe let's not assume `out/debug/` and, instead, assume that the directory we're passing to the `abigen!()` is the directory that contains both the `.bin` and `.json`? So a user would usually pass `/project_path/out/debug/`. That would relax the requirement of having to keep it all in `out/debug/`, but it would still require that both files are in the same directory. I _think_ that's okay?

**Comment by adlerjohn** [comment]
> I assume there are use cases where only the ABI is loaded without the binary.

Yes, since requiring the contract bytecode to interact with a contract defeats the purpose of a separate ABI file. The ABI alone should be sufficient to interact with a contract.

**Comment by digorithm** [comment]
> I assume there are use cases where only the ABI is loaded without the binary. By deployable mode I mean when the binary is loaded and the user wants to deploy it.
> 
> >

Perfect, that makes a lot of sense, thank you!

File: packages/fuels-contract/src/contract.rs (hunk 4 above)
Comment (MujkicA) [reply packages/fuels-contract/src/contract.rs:27]: `new`, other binding methods and any sway methods are generated within the same `impl` block. I ran a quick test with `new` defined in the json ABI to confirm we get a collision. I would therefore move new to the trait impl and keep it for now, maybe with a different name. If you know of a cleaner way to handle this, let me know.

File: packages/fuels-abigen-macro/src/lib.rs (hunk 6 above)
Comment (MujkicA) [reply packages/fuels-abigen-macro/src/lib.rs:110]: What would be the better way to do it or what is the problem with the way it is now?

File: packages/fuels-abigen-macro/tests/harness.rs (hunk 11 above)
Comment (vnepveu) [reply packages/fuels-abigen-macro/tests/harness.rs:977]: Since we are keeping the `load_sway` and `deploy` function, we should keep them tested. The question of removing those two functions can indeed be asked, but I think offering more granularity to the users is always nice, especially if the code is already there.

File: packages/fuels-core/src/code_gen/abigen.rs (hunk 10 above)
Comment (vnepveu) [reply packages/fuels-core/src/code_gen/abigen.rs:182]: summary of our discussion on Slack:
- The "onchain" workflow (ie loading the contract and deploying it) should not be dependent on the ABIgen workflow
- So we should just compress `load and deploy` into one function, but the contract instantiation `Contract::new()` should be separate. 

@digorithm  this is a decision that creates an extra "line of code" but IMO it's worth it compared to the confusion that could arise from having on-chain deployment being done in the ABIgen, what do you think?

File: packages/fuels-abigen-macro/src/lib.rs (hunk 6 above)
Comment (vnepveu) [reply packages/fuels-abigen-macro/src/lib.rs:110]: I would do a full branch with `!input.is_empty`, no? Since you're testing that condition twice?

**Comment by MujkicA** [comment]
Needs to be reopened

File: packages/fuels-abigen-macro/tests/harness.rs (hunk 11 above)
Comment (MujkicA) [reply packages/fuels-abigen-macro/tests/harness.rs]: Following the discussion on slack, I will create a separate issue to confirm if `load` has other use-cases. But I removed `deploy` to enforce a more streamlined approach via `load_and_deploy`.

File: packages/fuels-abigen-macro/tests/harness.rs [hunk 12]
```diff
@@ -1085,6 +1058,9 @@ async fn test_provider_launch_and_connect() {
         .unwrap();
     assert_eq!(42, result.value);
 
+    wallet.set_provider(launched_provider);
```
Comment (digorithm) [inline packages/fuels-abigen-macro/tests/harness.rs:1061]: I'm a bit confused about this test case now that we've made these changes. It feels like there's no difference between what all the other tests are doing; deploying is a different process than connecting to a deployed contract. Every test in here deploys and connects, and I'm not sure about the value of changing the provider here (`wallet.set_provider()`). Is this test case (`test_provider_launch_and_connect`) completely redundant now?

Comment (MujkicA) [reply packages/fuels-abigen-macro/tests/harness.rs:1061]: My understanding is that the test checks if behavior is consistent when making calls on a launched vs. connected `provider` instances. The binding now uses the `provider` from the supplied `wallet` to make method calls so I change it to `launched_provider` before making the second call.

Comment (digorithm) [reply packages/fuels-abigen-macro/tests/harness.rs:1061]: Fair enough, not entirely redundant then!

**Review by digorithm — CHANGES_REQUESTED** [review]
This is looking great -- much simpler! Also, extra nice because of the simple solution.

Just one more thing, can you update the [SDK book](https://github.com/FuelLabs/fuels-rs/blob/master/docs/src/getting-started/basics.md) to reflect the changes introduced here?

File: docs/src/getting-started/basics.md [hunk 13]
```diff
@@ -43,8 +43,8 @@ abigen!(
     "your_project/out/debug/contract_test-abi.json",
 );
 
-// This helper will setup a local node/provider and a test wallet.
-let (provider, wallet) = setup_test_provider_and_wallet().await;
+// This helper will setup a test wallet linked to a local node/provider.
```
Comment (vnepveu) [inline docs/src/getting-started/basics.md:46]: `will launch a local node and provide a test wallet linked to it`, I think it's important to mention that a local node gets launched by the function.

File: packages/fuels-abigen-macro/tests/harness.rs (hunk 12 above)
Comment (vnepveu) [reply packages/fuels-abigen-macro/tests/harness.rs:1060]: If this is the main test for testing `connected` vs `launched` instances, then the test `connect_to_deployed_contract` can be removed because it does this (actually badly now haha)

File: packages/fuels-abigen-macro/tests/harness.rs [hunk 14]
```diff
@@ -913,19 +899,17 @@ async fn test_connect_to_deployed_contract() {
         "packages/fuels-abigen-macro/tests/test_projects/contract_test/out/debug/contract_test-abi.json"
```
Comment (vnepveu) [inline packages/fuels-abigen-macro/tests/harness.rs:899]: remove this test (see comment below on `test_provider_launch_and_connect`)

File: packages/fuels-contract/src/contract.rs [hunk 15]
```diff
@@ -379,20 +379,18 @@ impl Contract {
             })
     }
 
-    /// Deploys a compiled contract to a running node
```
Comment (vnepveu) [inline packages/fuels-contract/src/contract.rs:382]: Don't remove the docstring

File: packages/fuels-test-helpers/src/signers.rs [hunk 16]
```diff
@@ -6,15 +6,14 @@ use fuels_signers::LocalWallet;
 use std::net::SocketAddr;
 
 #[cfg(feature = "fuels-signers")]
-pub async fn setup_test_provider_and_wallet() -> (Provider, LocalWallet) {
+pub async fn setup_test_provider_and_wallet() -> LocalWallet {
```
Comment (vnepveu) [inline packages/fuels-test-helpers/src/signers.rs:9]: Add a docstring

**Review by vnepveu — CHANGES_REQUESTED** [review]
This is looking supeeer clean now, I love it! Some documentation still needed

File: docs/src/getting-started/basics.md (hunk 13 above)
Comment (MujkicA) [reply docs/src/getting-started/basics.md:46]: I found it redundant since it is already stated at the beginning, but I guess it can't hurt to repeat it

File: packages/fuels-test-helpers/src/signers.rs (hunk 16 above)
Comment (digorithm) [reply packages/fuels-test-helpers/src/signers.rs:9]: Was this actually resolved? I'm not seeing the docstring @vnepveu suggested.

File: packages/fuels-contract/src/contract.rs (hunk 15 above)
Comment (digorithm) [reply packages/fuels-contract/src/contract.rs:382]: Was this actually resolved? I'm still seeing the docstring removed.

Comment (MujkicA) [reply packages/fuels-contract/src/contract.rs:382]: My bad

**Review by adlerjohn — COMMENTED** [review]
CI is failing due to some unrelated tests using the `master` branch for the stdlib. This should not be done; tests should use the implicit stdlib that matches the version of `forc` used in CI. Paging @vnepveu and @digorithm to investigate.

**Comment by vnepveu** [comment]
> CI is failing due to some unrelated tests using the `master` branch for the stdlib. This should not be done; tests should use the implicit stdlib that matches the version of `forc` used in CI. Paging @vnepveu and @digorithm to investigate.
Fixed in #280 (normally haha)

File: docs/src/getting-started/basics.md [hunk 17]
```diff
@@ -241,19 +241,19 @@ For a more concrete example, see the `test_contract_calling_contract` function i
 
 ## Connecting to existing contracts
 
-If you already have a deployed contract and want to call its methods using the SDK,  but without deploying it again, all you need is the contract ID of your deployed contract. You can skip the whole deployment setup and call `::new(contract_id, provider, wallet)` directly. For example:
+If you already have a deployed contract and want to call its methods using the SDK,  but without deploying it again, all you need is the contract ID of your deployed contract. You can skip the whole deployment setup and call `::new(contract_id, wallet)` directly. For example:
```
Comment (vnepveu) [inline docs/src/getting-started/basics.md:244]: Just realised that this example isn't accurate. If you have an already deplyed contract, then you already have a node. So you *have* to use a provider that connects to said node to connect the wallet to it. The instruction in this version would confuse users.

File: packages/fuels-contract/src/contract.rs [hunk 18]
```diff
@@ -380,19 +380,18 @@ impl Contract {
     }
 
     /// Deploys a compiled contract to a running node
-    /// To deploy a contract, you need a wallet with enough assets to pay for deployment. This
```
Comment (vnepveu) [inline packages/fuels-contract/src/contract.rs:383]: Re-add docstring

**Review by vnepveu — CHANGES_REQUESTED** [review]
One last detail I hadn't thought of then we merge :tada:

File: packages/fuels-test-helpers/src/signers.rs (hunk 16 above)
Comment (vnepveu) [reply packages/fuels-test-helpers/src/signers.rs]: This was not updated actually

File: docs/src/getting-started/basics.md (hunk 17 above)
Comment (MujkicA) [reply docs/src/getting-started/basics.md:244]: This was the case even before my changes. A workflow that involves connecting a provider to a node, instead of starting it via `launch_provider_and_get_wallet` is not introduced. I fear confusion might go the other way as well.

File: packages/fuels-test-helpers/src/signers.rs (hunk 16 above)
Comment (MujkicA) [reply packages/fuels-test-helpers/src/signers.rs]: Docstring added

Comment (adlerjohn) [reply packages/fuels-test-helpers/src/signers.rs]: https://github.com/MujkicA/contract-deployment-qol/commit/55964a5f8eee1b6e7f88c3ac3727d755806a0166#diff-b655abf20ad2575234c19c2517d33a0a4e681de106149f287130a67f1f96155eR8

File: docs/src/getting-started/basics.md (hunk 17 above)
Comment (vnepveu) [reply docs/src/getting-started/basics.md:244]: Can you please open a `docs` issue for this part then. This would be very confusing to a new user I think.

**Review by vnepveu — APPROVED** [review]
LGTM!!! :love_you_gesture:

**Review by digorithm — APPROVED** [review]
:shipit:
